Documentação: https://docs.scipy.org/doc/scipy/tutorial/stats/discrete.html

In [126]:
import scipy.stats
import numpy as np
import statistics

## Exercício 01

In [127]:
# Geralmente, quando não há reposição é Hipergeométrica
# Usa-se essa distribuição quando parece ser uma combinaçao

print(scipy.stats.hypergeom.pmf(1,195,9,5) * scipy.stats.hypergeom.pmf(1,200,10,5))

0.04039728314361033


## Exercício 02

Calculando a distribuição do número de caras:

In [128]:
# range(0,5) significa 0 caras, 1 cara, 2 caras, 3 caras ou 4 caras
# ou seja, em 4 lancamentos, podemos ter de 0 a 4 caras
print(scipy.stats.binom.pmf(range(0,5),4,3/8))

[0.15258789 0.36621094 0.32958984 0.13183594 0.01977539]


In [129]:
(3/8)**4

0.019775390625

In [130]:
# Px entao é a distribuicao do numero dec caras
Px = scipy.stats.binom.pmf(range(0,5),4,3/8)

Obtendo a distribuição de $W$:

In [131]:
# W é a multiplicacao de Caras e coroas
'''
0C 4K = 0
1C 3K = 3
2C 2K = 4
3C 1K = 3
4C 0K = 0

Px[i] = i caras
'''
# Sendo assim, montando a distribuiçao de probabilidades de W, usando as probabilidades de Px
# temos apenas 3 casos, já que w pode ser ou 0, ou 3 ou 4
Pw = [Px[0] + Px[4],Px[1] + Px[3],Px[2]]
print(Pw)

[0.17236328125, 0.49804687499999994, 0.3295898437500001]


Calculando $E[W]$: <br>
Obs: $E[W] = \Sigma_{i = 1}^{n} w_i \cdot p_i$

In [132]:
'''
Esperança de W pega os valores de w_i * p_i

w = 0 -> p = Pw[0]
w = 3 -> p = Pw[1]
w = 4 -> p = Pw[2]
'''
ew = 0*Pw[0] + 3*Pw[1] + 4*Pw[2]
print(ew)

2.8125


Calculando $Var[W]$ usando $E[W^2] - (E[W])^2$:

In [133]:
# No E[W^2], a probabilidade permanece a mesma, o que muda é apenas o w_i, que vira w_i**2
ew2 = 0*Pw[0] + 9*Pw[1] + 16*Pw[2]

In [134]:
varw = ew2 - ew**2
print(varw)

1.845703125


Calculando $Var[W]$ usando $E[(W - E[W])^2]$:

In [135]:
varbis = (0-ew)**2*Pw[0] + (3-ew)**2*Pw[1] + (4-ew)**2*Pw[2]
print(varbis)

1.8457031250000002


### Versao do Otavio

In [136]:
# num de caras multiplicado pelo num de coroas

# binomial 

# 0 1 2 3 4 caras

# 0 3 4 3 0 valores possiveis
probs = []
for i in range (0,5):
  probs.append(scipy.stats.binom.pmf(i,4,0.375))  # (primeiro arg é a variavel discreta, a segunda a qauntitade de amostras e a terceira a probabilidade) acumulada

print(probs)

# calculando a esperança
hope = 0
values_poss = [0,3,4,3,0]
for i in range (0,5):
  hope += values_poss[i]*probs[i]


# calcular variancia

var = 0
values_poss = [0,3,4,3,0]
for i in range (0,5):
  var += ((values_poss[i] - hope)**2)*probs[i]


# encontrando a mediana
# Construindo uma distribuicao de probabilidade acumulada, ou seja:
''' 
0C 4K = 0
1C 3K = 3
2C 2K = 4
3C 1K = 3
4C 0K = 0

w = 0 -> probs[0] + probs[4]
w = 3 -> probs[1] + probs[3]
w = 4 -> probs[2]
'''
# Queremos construir uma funcao de distribuicao acumulada, para saber em qual w_i a probabilidade passa de 50%
values_poss = [0, 3, 4]
probs[0] = probs[0] + probs[4]
probs[1] = probs[0] + probs[1] + probs[3]
probs[2] = probs[1] + probs[2]
del probs[3:]
'''
w = 0 -> p = 0.1
w = 3 -> p = 0.4
w = 4 -> p = 1
'''

# Aqui eu encontro qual a mediana do meu vetor de probabilidade, e encontro qual índice dessa mediana
median_index = (probs.index(statistics.median(probs)))
# Agora, pegaremos esse índice e printaremos a mediana, usando a lista values_poss
print(values_poss[median_index])
# a mediana é a var disceta que tem a probabilidade acumulada >= 50%

[0.152587890625, 0.3662109375, 0.3295898437500001, 0.13183593749999994, 0.019775390625]
3


## Exercício 03

Calculando a esperança:

In [137]:
# para um componente: 10 falhas a cada 300 horas = 10/300 falhas/hora, ou 0.033333
# para cinco componentes: 5 * 10/300 falhas/hora
# para 75 horas: 75 * 5 * 10/300 falhas = 12.5 falhas em 75 horas para 5 componentes
# portanto a esperança é 12.5 falhas

Alguns valores de da distribuição $W$:

In [139]:
Pw = scipy.stats.poisson.pmf(range(0,50),12.5)
print(Pw)

[3.72665317e-06 4.65831647e-05 2.91144779e-04 1.21310325e-03
 3.79094764e-03 9.47736911e-03 1.97445190e-02 3.52580696e-02
 5.50907338e-02 7.65149080e-02 9.56436350e-02 1.08685949e-01
 1.13214530e-01 1.08860125e-01 9.71965402e-02 8.09971169e-02
 6.32789975e-02 4.65286747e-02 3.23115796e-02 2.12576182e-02
 1.32860114e-02 7.90834010e-03 4.49337505e-03 2.44205166e-03
 1.27190191e-03 6.35950953e-04 3.05745651e-04 1.41548912e-04
 6.31914787e-05 2.72377063e-05 1.13490443e-05 4.57622754e-06
 1.78758888e-06 6.77117002e-07 2.48940074e-07 8.89071693e-08
 3.08705449e-08 1.04292381e-08 3.43067044e-09 1.09957386e-09
 3.43616831e-10 1.04761229e-10 3.11789372e-11 9.06364454e-12
 2.57489902e-12 7.15249727e-13 1.94361339e-13 5.16918454e-14
 1.34614181e-14 3.43403522e-15]


Uma intuição de que realmente $E[W] = λ_W$:

In [140]:
ew = 0
for w in range(0,50):
    ew += w * Pw[w]
print(ew)

12.499999999999972


Uma intuição de que realmente $Var[W] = λ_W$:

In [141]:
varw = 0
for w in range(0,50):
    varw += (w - ew)**2 * Pw[w]
print(varw)

12.499999999998394


Valor mais provável de W:

In [162]:
# Já que a esperança de W é 12.5, seu valor mais provável é 12
# P(W = 12):
# Obs: Usando pmf eu sei a probabilidade quando W = 12, usando cdf eu sei a probabilidade de W <= 12
scipy.stats.poisson.pmf(12,12.5)


0.11321453006756854

Dsitribuição acumulada de $W$:

In [160]:
cumulative_poisson = scipy.stats.poisson.cdf(range(0,50),12.5)
# Isso printa com 3 casas decimais
np.set_printoptions(formatter={'float': lambda x: "{0:0.3f}".format(x)})
print(cumulative_poisson)

[0.000 0.000 0.000 0.002 0.005 0.015 0.035 0.070 0.125 0.201 0.297 0.406
 0.519 0.628 0.725 0.806 0.869 0.916 0.948 0.969 0.983 0.991 0.995 0.998
 0.999 0.999 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000
 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000 1.000
 1.000 1.000]


Acima é preciso encontrar o primeiro valor maior do que $0.5$ para saber a mediana. Nesse caso, o $12^o$ valor é o primeiro valor acima de 0.5, que é 0.519

In [158]:
for i, j in zip(cumulative_poisson, range(50)):
    if i > 0.5:
        print(j)
        break

12


Abaixo a mediana sai mais rápido:

In [143]:
scipy.stats.poisson.ppf(0.5,12.5)

12.0

## Exercício 04

In [144]:
print(scipy.stats.poisson.pmf(0,1.4) * scipy.stats.poisson.pmf(0,1.12))

0.08045960674953244


In [145]:
print(np.exp(-2.52))

0.08045960674953244


In [146]:
print(scipy.stats.poisson.pmf(2,1.4) * scipy.stats.poisson.pmf(1,1.12))

0.0883124643682868


### Turma 01:

Vendo os primeiros valores da somatória:

In [147]:
print((1 - scipy.stats.poisson.cdf(0,1.4)) * scipy.stats.poisson.pmf(0,1.12))

0.24582018787350704


In [148]:
print((1 - scipy.stats.poisson.cdf(1,1.4)) * scipy.stats.poisson.pmf(1,1.12))

0.14915794703506097


In [149]:
print((1 - scipy.stats.poisson.cdf(2,1.4)) * scipy.stats.poisson.pmf(2,1.12))

0.03407347029339359


In [150]:
print((1 - scipy.stats.poisson.cdf(3,1.4)) * scipy.stats.poisson.pmf(3,1.12))

0.004104605719255237


Fazendo aproximação da somatória:

In [151]:
soma = 0
for w in range(0,50):
    soma += (1 - scipy.stats.poisson.cdf(w,1.4)) * scipy.stats.poisson.pmf(w,1.12)
print(soma)

0.4334770286856959


### Turma 02

Vendo os primeiros valores da somatória:

In [152]:
print(scipy.stats.poisson.cdf(0,1.12) * scipy.stats.poisson.pmf(1,1.4))

0.11264344944934539


In [153]:
print(scipy.stats.poisson.cdf(1,1.12) * scipy.stats.poisson.pmf(2,1.4))

0.16716287898282858


In [154]:
print(scipy.stats.poisson.cdf(2,1.12) * scipy.stats.poisson.pmf(3,1.4))

0.10108833421356564


Fazendo aproximação da somatória:

In [155]:
p = 0
for w in range(0,50):
    p += scipy.stats.poisson.cdf(w,1.12) * scipy.stats.poisson.pmf(w+1,1.4)
print(p)

0.43347702868569593
